Add governance policies (#38)
cdevents/community: processes.md

Reviewer: Should we include an initial list of people in all these roles as part of this PR?

Author: I would prefer to keep the list documented in their respective places.
Some of the lists are maintained in GitHub, e.g. org members are listed in https://github.com/orgs/cdevents/people. Governing board members are documented in governance.md.

# CDEvents project processes

This doc explains general development processes that apply to all projects within the org. Individual projects may have their own processes as well, which you can find documented in their individual CONTRIBUTING.md files.

## Contributions

We try and track community contributions as much as possible to measure the health of the project.

> Contributions can include opening PRs, reviewing and commenting on PRs, opening and commenting on issues, writing design docs, commenting on design docs, helping people on slack, and participating in working groups. Where possible, we use dashboards on cdevents.devstats.cd.foundation to track measurable engagement. We try our best to include contributions that are not GitHub, but accuracy varies when we don't have easily available data.

Reviewer: ```suggestion
Contributions can include opening PRs, reviewing and commenting on PRs, opening and commenting on issues, writing design docs, commenting on design docs, helping people on slack, and participating in working groups. Where possible, we use dashboards on [cdevents.devstats.cd.foundation](https://cdevents.devstats.cd.foundation) to track measurable engagement. We try our best to include contributions that are not GitHub, but accuracy varies when we don't have easily available data.
```

### Organization Member

Description: An Organization Member is an established contributor who regularly participates in the project. Organization Members have privileges in project repositories.

An Organization Member must meet the responsibilities and has the requirements of a [Contributor](#contributor), plus:

* Responsibilities include:
  * Continues to contribute regularly, as demonstrated by having at least 15
    contributions a year, as demonstrated by
    [the CDEvents devstats dashboard](https://cdevents.devstats.cd.foundation/d/9/developer-activity-counts-by-repository-group-table?orgId=1&var-period_name=Last%20year&var-metric=contributions&var-repogroup_name=All&var-country_name=All).
* Requirements:
  * Subscribed to the [cdevents-dev mailing list](/README.md#how-to-get-involved).
  * Actively contributing to 1 or more subprojects.
  * Must have successful contributions to the project, including at least one of
    the following:
    * Authored or Reviewed 5 PRs,
    * Or be endorsed by at least 2 existing Org Members:
      * Endorsers should have close interactions with the prospective member - e.g. code/design/proposal review, coordinating on issues, etc.

>       * Endorsers must be maintainers of at lease one of the CDEvents sub-projects.

Reviewer: ```suggestion
      * Endorsers must be maintainers of at least one of the CDEvents sub-projects.
```

      * Endorsers must be from multiple member organizations to demonstrate integration across community.
  * Must have 2FA enabled on their GitHub account
* Privileges:
  * [GitHub default organization member privileges](https://docs.github.com/en/organizations/managing-peoples-access-to-your-organization-with-roles/roles-in-an-organization#permissions-for-organization-roles)
  * Can recommend other contributors to become Org Members

The process for a Contributor to become an Organization Member is as follows:

1. Request an organization owner to be added to the GitHub organization

## Maintainer

Description: Maintainers are very established contributors who are responsible for entire projects. As such, they have the ability to approve PRs against any area of a project, and are expected to participate in making decisions about the strategy and priorities of the project.

A Maintainer must meet the responsibilities and requirements of an [organization member](#organization-member), plus:

- Responsibilities include:
  - Proactively help triage and respond to incoming issues (GitHub, Slack, mailing list)
  - Reviewing PRs that involve multiple parts of the project

>   - Mentoring new [Contributors](#contributor) and [Reviewers](#Reviewer)

Reviewer: ```suggestion
  - Mentoring new [Contributors](#contributor)
```

  - Writing PRs that involve many parts of the project (e.g. refactoring)
  - Participating in CDEvents maintainer activities (Facilitate working group, triage issues)
  - Determining strategy and policy for the project
  - Participating in, and leading, community meetings
  - Mentoring other contributors
- Requirements

>   - Have been actively participating in reviews for at least 3 months or 50% of the project lifetime, whichever is shorter

Reviewer: ```suggestion
  - Has been actively participating in reviews for at least 3 months or 50% of the project lifetime, whichever is shorter
```

  - Has authored or reviewed at least 10 PRs to the codebase.

>     - Have been the primary reviewer for at least 5 substantial PRs to the codebase.

Reviewer: ```suggestion
    - Has been the primary reviewer for at least 5 substantial PRs to the codebase.
```

  - Failing to meet role requirements.
  - Periods of no [contributions](./README.md#contributions) for longer than 4 months
  - Periods of no communication for longer than 2 months
- Consequences of being inactive include:
  - Involuntary removal or demotion
  - Being asked to move to Emeritus status

### Involuntary Removal or Demotion

Involuntary removal/demotion of a contributor (at any level) happens when responsibilities and requirements aren't being met. This may include repeated patterns of inactivity, extended period of inactivity, a period of failing to meet the requirements of your role, and/or a violation of the Code of Conduct. This process is important because it protects the community and its deliverables while also opens up opportunities for new contributors to step in.

Involuntary removal or demotion is handled through a vote by a majority of the [CDEvents Governing Board](/governance.md).

### Stepping Down/Emeritus Process

If and when contributors' commitment levels change, contributors (at any level) can consider stepping down (moving down the contributor ladder) vs moving to emeritus status (completely stepping away from the project).

> [Community participants](#community-participant), [contributors](#contributor) and [organization members](#organization-member) may change their contribution level autonomously.

Reviewer: ```suggestion
[Community participants](#community-participant), [contributors](#contributor) and [organization members](#organization-member) may reduce their contributor level autonomously.
```

### Organization Member

Description: An Organization Member is an established contributor who regularly participates in the project. Organization Members have privileges in project repositories.

An Organization Member must meet the responsibilities and has the requirements of a [Contributor](#contributor), plus:

* Responsibilities include:
  * Continues to contribute regularly, as demonstrated by having at least 15
    contributions a year, as demonstrated by
    [the CDEvents devstats dashboard](https://cdevents.devstats.cd.foundation/d/9/developer-activity-counts-by-repository-group-table?orgId=1&var-period_name=Last%20year&var-metric=contributions&var-repogroup_name=All&var-country_name=All).
* Requirements:
  * Subscribed to the [cdevents-dev mailing list](/README.md#how-to-get-involved).
  * Actively contributing to 1 or more subprojects.
  * Must have successful contributions to the project, including at least one of
    the following:
    * Authored or Reviewed 5 PRs,
    * Or be endorsed by at least 2 existing Org Members:
      * Endorsers should have close interactions with the prospective member - e.g. code/design/proposal review, coordinating on issues, etc.

>       * Endorsers must be reviewers or approvers in at least one OWNERS file within one of the Tekton sub-projects.

Reviewer 2: ```suggestion
      * Endorsers must be reviewers or approvers in at least one OWNERS file within one of the CDEvents sub-projects.
```

Author: Fixed

* Or, Be nominated by another OWNER (with no objections from other OWNERs)

The process of becoming a Reviewer is:

1. The contributor is nominated by opening a PR against the appropriate project, adding the contributor to `REVIEWERS.md`
2. At least two Reviewers/[Maintainers](#maintainer) of the team that owns that repository approve the nomination PR.
3. Update the GitHub project reviewer team

* Each project has a `<repo>.-reviewers` team defined, where `<repo>` is the name of the GitHub repository.

## Maintainer

Description: Maintainers are very established contributors who are responsible for entire projects. As such, they have the ability to approve PRs against any area of a project, and are expected to participate in making decisions about the strategy and priorities of the project.

A Maintainer must meet the responsibilities and requirements of a [Reviewer](#Reviewer), plus:

- Responsibilities include:
  - Reviewing PRs that involve multiple parts of the project
  - Mentoring new [Contributors](#contributor) and [Reviewers](#Reviewer)
  - Writing PRs that involve many parts of the project (e.g. refactoring)

>   - Participating in Tekton maintainer activities (build captain, WG lead)

Reviewer 2: ```suggestion
  - Participating in CDEvents maintainer activities (build captain, WG lead)
```

  - Determining strategy and policy for the project
  - Participating in, and leading, community meetings
  - Mentoring other contributors
- Requirements
  - Have been actively participating in reviews for at least 3 months or 50% of the project lifetime, whichever is shorter
  - Has authored or reviewed at least 10 PRs to the codebase.
    - Have been the primary reviewer for at least 5 substantial PRs to the codebase.
  - Demonstrates a broad knowledge of the project across multiple areas
  - Is able to exercise judgement for the good of the project, independent of their employer, friends, or team
  - Be nominated by another Maintainer (with no objections from other Maintainers)
- Additional privileges:
  - Approve PRs to any area of the project

>   - Granted access to shared Tekton CI/CD infrastructure

Reviewer 2: ```suggestion
  - Granted access to shared CDEvents CI/CD infrastructure
```

Author: Thanks, good catch

- Inactivity is measured by:
  - Failing to meet role requirements.
  - Periods of no [contributions](./README.md#contributions) for longer than 4 months
  - Periods of no communication for longer than 2 months
- Consequences of being inactive include:
  - Involuntary removal or demotion
  - Being asked to move to Emeritus status

### Involuntary Removal or Demotion

Involuntary removal/demotion of a contributor happens when responsibilities and requirements aren't being met. This may include repeated patterns of inactivity, extended period of inactivity, a period of failing to meet the requirements of your role, and/or a violation of the Code of Conduct. This process is important because it protects the community and its deliverables while also opens up opportunities for new contributors to step in.

Involuntary removal or demotion is handled through a vote by a majority of the [CDEvents Governing Board](/governance.md).

### Stepping Down/Emeritus Process

If and when contributors' commitment levels change, contributors can consider stepping down (moving down the contributor ladder) vs moving to emeritus status (completely stepping away from the project).

> Contact the Maintainers about changing to Emeritus status, or reducing your contributor level.

Reviewer 3: Can one become an Emeritus and then come back? If so do we need to write something about it?

Reviewer 3: But I was referring to the contribution ladder not the governing board.

  * Submitting bug reports
  * Commenting on issues
  * Trying out new releases
  * Attending community events

### Contributor

Description: A Contributor makes direct contributions to the project and adds value to it. [Contributions need not be code](#contributions). People at the Contributor level may be new contributors, and they can contribute occasionally.

Contributors may be eligible to vote and run in elections. See [Elections](/governance.md#elections) for more details.

A Contributor must meet the responsibilities of a [Community Participant](#community-participant), plus:

* Responsibilities include:
  * Follow the project contributing guide (see `CONTRIBUTING.md` in the
    corresponding repo)
* Requirements (one or several of the below):
  * Report and sometimes resolve issues
  * Occasionally submit PRs
  * Contribute to the documentation
  * Participate in [meetings](/README.md#how-to-get-involved)
  * Answer questions from other community members
  * Submit feedback on issues and PRs
  * Test, review, and verify releases and patches
* Privileges:
  * Invitations to contributor events
  * Eligible to become an [Organization Member](#organization-member)

> 

Reviewer: What is the process to become a Contributor? Where are contributors kept track of?

Author: There is no "process" for becoming a contributor. Whoever contributes to CDEvents in various forms is a contributor. 

There is no easy way to keep track of contributors either, unfortunately. 
One way we achieve that is through devstats, but that does not necessarily cover everything. 
Asking people to register or fill out a form before making a contribution would raise the bar to contributions, and that's not something we want.

In case of elections though, we may ask people to fill a form to declare themselves as contributors if they want to vote and are not tracked by devstat

Reviewer: That's all good. I guess I was considering e.g. the privilege of being invited to contributor events (whatever that is), since then we would need a well-defined list of people with contact info to be able to reach them. But I guess we could copy that from devstats. I'm glad we don't add extra admin to manage the list of contributors manually.

* Requirements:
  * Subscribed to the [cdevents-dev mailing list](/README.md#how-to-get-involved).
  * Actively contributing to 1 or more subprojects.
  * Must have successful contributions to the project, including at least one of
    the following:
    * Authored or Reviewed 5 PRs,
    * Or be endorsed by at least 2 existing Org Members:
      * Endorsers should have close interactions with the prospective member - e.g. code/design/proposal review, coordinating on issues, etc.
      * Endorsers must be reviewers or approvers in at least one OWNERS file within one of the CDEvents sub-projects.
      * Endorsers must be from multiple member organizations to demonstrate integration across community.
  * Must have 2FA enabled on their GitHub account
* Privileges:
  * [GitHub default organization member privileges](https://docs.github.com/en/organizations/managing-peoples-access-to-your-organization-with-roles/roles-in-an-organization#permissions-for-organization-roles)
  * Can recommend other contributors to become Org Members

The process for a Contributor to become an Organization Member is as follows:

1. Request an organization owner to be added to the GitHub organization

> ## Reviewer

Author: Keep or remove?

  - Be nominated by another Maintainer (with no objections from other Maintainers)
- Additional privileges:
  - Approve PRs to any area of the project
  - Granted access to shared CDEvents CI/CD infrastructure
  - Represent the project in public as a Maintainer
  - Have a vote in Maintainer decision-making meetings

To facilitate productivity, small repositories, repositories that do not contain production code or that have a low traffic of PRs may decide to use simpler requirements. Each of these projects may define their own requirements, in a local `processes.md` file.

Process of becoming an Maintainer:

1. Any current Maintainer may nominate a current [Reviewer](#Reviewer) to become a new Maintainer, by opening a PR against the appropriate project, adding the contributor to `MAINTAINERS.md`
2. The nominee will add a comment to the PR testifying that they agree to all requirements of becoming a Maintainer.
3. A majority of the current Maintainers must then approve.
4. Add the new maintainer to the corresponding GitHub team

- Each project has a `<repo>-maintainers` team, where `<repo>` is the name of the GitHub repository.

> ## Governance Committee Member

Reviewer: ```suggestion
## Governing Board Member
```

Author: Fixed, thank you

> Description: The CDEvents Governance committee is the governing body of the CDEvents open source project. It's an elected group that represents the contributors to the project, and has an oversight on governance and technical matters.

Reviewer: ```suggestion
Description: The CDEvents Governing Board is the governing body of the CDEvents open source project. It's an elected group that represents the contributors to the project, and has an oversight on governance and technical matters.
```

Author: Fixed, thank you

See [governance.md](/governance.md) for requirements, responsibilities, and election process.

- Additional privileges:
  - Maintainer privileges on all CDEvents projects
  - Organization admin access.

> Members of the governance committee belong to the the `governance` GitHub team.

Reviewer: ```suggestion
Members of the Governing Board belong to the the `governance` GitHub team.
```

Author: Thanks

- Inactivity is measured by:
  - Failing to meet role requirements.
  - Periods of no [contributions](./README.md#contributions) for longer than 4 months
  - Periods of no communication for longer than 2 months
- Consequences of being inactive include:
  - Involuntary removal or demotion
  - Being asked to move to Emeritus status

### Involuntary Removal or Demotion

Involuntary removal/demotion of a contributor happens when responsibilities and requirements aren't being met. This may include repeated patterns of inactivity, extended period of inactivity, a period of failing to meet the requirements of your role, and/or a violation of the Code of Conduct. This process is important because it protects the community and its deliverables while also opens up opportunities for new contributors to step in.

Involuntary removal or demotion is handled through a vote by a majority of the [CDEvents Governing Board](/governance.md).

### Stepping Down/Emeritus Process

> If and when contributors' commitment levels change, contributors can consider stepping down (moving down the contributor ladder) vs moving to emeritus status (completely stepping away from the project).

Author: s/contributor/maintainer
